### PR TITLE
make targets active parallel

### LINKED ISSUE #5717: Make Manager.TargetsActive parallel
Currently https://github.com/prometheus/prometheus/blob/32c3a1beeffbcf0d308f724931d8ed46e8048018/scrape/manager.go#L222 goes in series through the scrape pools. If there's an update from SD, this can end up blocking on each scrapePool's `Sync` in turn. Doing this in parallel would limit this to the scrape interval.

### PR DESCRIPTION
**Pull request by yeya24**
Signed-off-by: yeya24 <yb532204897@gmail.com>

fix #5717

### CODE REVIEW

File: scrape/manager.go [hunk 1]
```diff
@@ -223,10 +223,18 @@ func (m *Manager) TargetsActive() map[string][]*Target {
 	m.mtxScrape.Lock()
 	defer m.mtxScrape.Unlock()
 
+	var wg sync.WaitGroup
 	targets := make(map[string][]*Target, len(m.scrapePools))
 	for tset, sp := range m.scrapePools {
-		targets[tset] = sp.ActiveTargets()
+		wg.Add(1)
+		// Run in parallel limit the blocking time of scrapePool to scrape
```
Comment (codesome) [inline scrape/manager.go]:
```suggestion
		// Runing in parallel limits the blocking time of scrapePool to scrape
```

File: scrape/manager.go [hunk 2]
```diff
@@ -223,10 +223,18 @@ func (m *Manager) TargetsActive() map[string][]*Target {
 	m.mtxScrape.Lock()
 	defer m.mtxScrape.Unlock()
 
+	var wg sync.WaitGroup
 	targets := make(map[string][]*Target, len(m.scrapePools))
 	for tset, sp := range m.scrapePools {
-		targets[tset] = sp.ActiveTargets()
+		wg.Add(1)
```
Comment (codesome) [inline scrape/manager.go]: You can have a single `wg.Add(len(m.scrapePools))` call before the `for` loop.

File: scrape/manager.go (hunk 1 above)
Comment (codesome) [reply scrape/manager.go]: `Runing` -> `Running`

File: scrape/manager.go (hunk 2 above)
Comment (yeya24) [reply scrape/manager.go]: Done.

File: scrape/manager.go (hunk 1 above)
Comment (yeya24) [reply scrape/manager.go]: Done.

File: scrape/manager.go [hunk 3]
```diff
@@ -223,10 +223,18 @@ func (m *Manager) TargetsActive() map[string][]*Target {
...
 
+	var wg sync.WaitGroup
 	targets := make(map[string][]*Target, len(m.scrapePools))
+	wg.Add(len(m.scrapePools))
 	for tset, sp := range m.scrapePools {
-		targets[tset] = sp.ActiveTargets()
+		// Running in parallel limits the blocking time of scrapePool to scrape
+		// interval when there's an update from SD.
+		go func(tset string, sp *scrapePool) {
+			targets[tset] = sp.ActiveTargets()
```
Comment (brian-brazil) [inline scrape/manager.go]: Do we need some locking here?

Comment (brian-brazil) [reply scrape/manager.go]: Did you have a chance to look at this?

Comment (yeya24) [reply scrape/manager.go]: Yes, I will finish it  soon.

File: scrape/manager.go [hunk 4]
```diff
@@ -61,6 +61,7 @@ type Manager struct {
 	graceShut chan struct{}
 
 	jitterSeed    uint64     // Global jitterSeed seed is used to spread scrape workload across HA setup.
+	mtxTargets    sync.Mutex // Guards for active targets.
```
Comment (brian-brazil) [inline scrape/manager.go]: This should be a local variable, it's only protecting a local variable.

File: scrape/manager.go (hunk 3 above)
Comment (sylr) [reply scrape/manager.go]: I might be wrong but isn't this a case of https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables ?

```
go func(tset string, sp *scrapePool) {
 	...
}(tset, &m.scrapePools[tset])
```

https://play.golang.org/p/FTB20cdWLXu

Comment (brian-brazil) [reply scrape/manager.go]: The code looks fine to me

Comment (sylr) [reply scrape/manager.go]: I think we do need locking. As far as I understand you can't have several goroutines writing in the same map at the same time (even if they modify different keys of the map).

Comment (brian-brazil) [reply scrape/manager.go]: The locking has already been added

Comment (sylr) [reply scrape/manager.go]: > The code looks fine to me

It looks fine but the goroutines are stacked and can be started after the end of the iteration of the `for loop` (especially if there is nothing else in the for loop body that the goroutine calls).

If you pass a pointer as argument of the goroutine function closure you might end up on calling all the goroutines with the same pointer, the one of the last element of the map you iterate on.

See the following example: https://play.golang.org/p/FTB20cdWLXu

Comment (brian-brazil) [reply scrape/manager.go]: We're not passing `&sp` though, we're passing `sp`

Comment (sylr) [reply scrape/manager.go]: Indeed, it works in this case: https://play.golang.org/p/XuBiUEu59ci

**Comment by brian-brazil** [comment]
Thanks!

File: scrape/manager.go [hunk 5]
```diff
@@ -223,10 +222,24 @@ func (m *Manager) TargetsActive() map[string][]*Target {
...
+	)
+
 	targets := make(map[string][]*Target, len(m.scrapePools))
+	wg.Add(len(m.scrapePools))
 	for tset, sp := range m.scrapePools {
-		targets[tset] = sp.ActiveTargets()
+		// Running in parallel limits the blocking time of scrapePool to scrape
+		// interval when there's an update from SD.
+		go func(tset string, sp *scrapePool) {
+			mtx.Lock()
```
Comment (bboreham) [inline scrape/manager.go:236]: Does this really run in parallel?  As far as I can see it's all serialized by the mutex - what did you gain?

**Pull request merged**